File: common/src/main/java/dev/obscuria/tooltips/client/filter/RarityFilter.java
package dev.obscuria.tooltips.client.filter;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.world.item.ItemStack;

public record RarityFilter(String rarity) implements ItemFilter {

    public static final Codec<RarityFilter> CODEC;

    @Override
    public Codec<RarityFilter> codec() {
        return CODEC;
    }

    @Override
    public boolean test(ItemStack stack) {
        final var rarity = stack.getRarity().name().toLowerCase();
        return this.rarity.equals(rarity);
    }

    static {
        CODEC = RecordCodecBuilder.create(codec -> codec.group(
                Codec.STRING.fieldOf("rarity").forGetter(RarityFilter::rarity)
        ).apply(codec, RarityFilter::new));
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/filter/RarityFilter.kt
package dev.obscuria.tooltips.client.filter

import com.mojang.serialization.Codec
import com.mojang.serialization.DataResult
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Rarity
import java.util.Optional

class RarityFilter(
    val rarity: Rarity
): IItemFilter {

    override fun codec(): Codec<RarityFilter> = CODEC

    override fun test(stack: ItemStack): Boolean {
        return stack.rarity.equals(rarity)
    }

    companion object {

        private val RARITY_CODEC: Codec<Rarity> = Codec.STRING.comapFlatMap(::rarityByName, ::nameOfRarity)
        val CODEC: Codec<RarityFilter> = RecordCodecBuilder.create { codec ->
            codec.group(
                RARITY_CODEC
                    .fieldOf("rarity")
                    .forGetter(RarityFilter::rarity)
            ).apply(codec, ::RarityFilter)
        }

        private fun rarityByName(name: String): DataResult<Rarity> {
            return Optional.ofNullable(Rarity.entries.firstOrNull { it.name.lowercase() == name.lowercase() })
                .map { DataResult.success(it) }
                .orElseGet { DataResult.error { "Unknown rarity: $name" } }
        }

        private fun nameOfRarity(rarity: Rarity): String {
            return rarity.name.lowercase()
        }
    }
}